### PR TITLE
[Metrics] Added Sensitivity and Specificity. Added pos_labels param.

### PR DESCRIPTION
**Pull request by shijianjian**
## What does this PR do?

- Added Sensitivity and Specificity (useful for medical projects).
- Added pos_labels param (#4791).

See if you are interested. If so, I will add tests and the corresponding doc updates.

Fixes # (issue)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/metrics/utils.py [hunk 1]
```diff
@@ -146,3 +146,35 @@ def _input_format_classification_one_hot(
...
+        >>> one_hot_labels = torch.tensor([
+        ...    [1, 0, 0, 1, 1, 0],
+        ...    [0, 0, 1, 0, 0, 1],
+        ...    [0, 1, 0, 0, 0, 0]
+        ... ])  # num_class * n (3 class, 6 examples)
+        >>> pos_labels = [1]
+        >>> _input_one_hot_filtering(one_hot_labels, pos_labels)
+        tensor([[1, 0, 0, 1, 1, 0],
+                [0, 0, 0, 0, 0, 0],
+                [0, 1, 0, 0, 0, 0]])
```
Comment (Borda) [inline pytorch_lightning/metrics/utils.py:171]: if I understand "Filter out" as drop non-positive, then you have it here swapped bc you show only the non-positive

File: pytorch_lightning/metrics/classification/precision_recall.py [hunk 2]
```diff
@@ -218,3 +237,102 @@ def compute(self):
             return self.true_positives.sum().float() / (self.actual_positives.sum() + METRIC_EPS)
         elif self.average == 'macro':
             return (self.true_positives.float() / (self.actual_positives + METRIC_EPS)).mean()
+
+
+Sensitivity = Recall
+
+
+class Specificity(Metric):
```
Comment (Borda) [inline pytorch_lightning/metrics/classification/precision_recall.py:245]: can we define it as functional and then add class as a wrapper?

File: pytorch_lightning/metrics/utils.py (hunk 1 above)
Comment (shijianjian) [reply pytorch_lightning/metrics/utils.py:171]: I will do a renaming to make it more clearer.

File: pytorch_lightning/metrics/classification/precision_recall.py (hunk 2 above)
Comment (shijianjian) [reply pytorch_lightning/metrics/classification/precision_recall.py:245]: I could add the functional. 

While for the modulized parts, I followed the same style as Recall and Precision. If a wrapper shall be used, should I update those functions as well?

Comment (SkafteNicki) [reply pytorch_lightning/metrics/classification/precision_recall.py:349]: @shijianjian then intention is that all class metrics should call their functional counterpart, but recall and precision have still not been updated to this format. Please see how the `ConfusionMatrix` metric is organised :]

**Comment by Borda** [comment]
@shijianjian mind resolve conflicts so can let it land? :]
cc: @SkafteNicki @justusschock

File: pytorch_lightning/metrics/functional/classification.py [hunk 3]
```diff
@@ -104,9 +107,31 @@ def stat_scores_multiple_classes(
...
+        tensor([2.])
+
     """
     if pred.ndim == target.ndim + 1:
+        # Convert onehot pred tensor to categorical
         pred = to_categorical(pred, argmax_dim=argmax_dim)
+    elif pred.ndim == target.ndim == 1:
+        # Convert numerical pred tensor to categorical
+        if torch.is_floating_point(pred):
+            pred = pred.round().long()
```
Comment (shijianjian) [inline pytorch_lightning/metrics/functional/classification.py:133]: I think I am facing a precision error posed by ```torch.round```

Comment (Borda) [reply pytorch_lightning/metrics/functional/classification.py:133]: maybe you can separate this functionality as an extra function and add some code sample aka doctest to help debug what is happening there

**Comment by hassiahk** [comment]
@shijianjian, aren't sensitivity and recall same? or is there any difference when it comes to medical applications?

**Comment by shijianjian** [comment]
> @shijianjian, aren't sensitivity and recall same? or is there any difference when it comes to medical applications?

Yes, they are the same. So that the sensitivity is an alias to recall in the code.

If lightning does not mean to make all alias available, I think it is better to use naming system like TPR, TNR instead.

**Comment by hassiahk** [comment]
@shijianjian I don't think aliases are necessary since in the docs it is mentioned how a specific metric is calculated.

**Comment by SkafteNicki** [comment]
@shijianjian what about updating the documentation of `recall` such that it states something like: "...also known as sensitivity or true positive rate..."?

**Comment by shijianjian** [comment]
> @shijianjian what about updating the documentation of `recall` such that it states something like: "...also known as sensitivity or true positive rate..."?

Yes, we can update the doc for that. 

If so, my suggestion is to use TPR, TNR, PPV instead of Recall, Specificity and Precision since they are not in the same naming system. I feel like we are mixing concepts up.

**Comment by hassiahk** [comment]
In my personal opinion, most of the people would know the terms `precision`, `recall`, `specificity` better than `PPV`, `TPR`, `TNR`. So, I think it would be better to add those terms to the documentation like @SkafteNicki said.

**Comment by shijianjian** [comment]
Specificity is essentially Recall with ```pos_label==0```.

If you want to remove the alias, I think specificity can also be removed in that case.  The only problem is that it will not make sense semantically since it will count ```true_positives``` and ```conditional_positives```  internally, which might confuse some users.

**Comment by SkafteNicki** [comment]
@shijianjian so is the fix to include the `pos_label` arg in `Recall` and then it can do everything that the sensitivity and specificity metrics you have also included?

**Comment by shijianjian** [comment]
> @shijianjian so is the fix to include the `pos_label` arg in `Recall` and then it can do everything that the sensitivity and specificity metrics you have also included?

Yes. To maximumly reduce the complexity, from the functionality perspective, a recall with pos_label would be sufficient. 

Though I would suggest to have ~~~Sensitivity = Recall~~~ and 

~~~python 
Class Specificity(Recall):
    def __init__(self, ..., pos_label=0):
        ...
~~~

**Comment by SkafteNicki** [comment]
> 
> 
> > @shijianjian so is the fix to include the `pos_label` arg in `Recall` and then it can do everything that the sensitivity and specificity metrics you have also included?
> 
> Yes. To maximumly reduce the complexity, from the functionality perspective, a recall with pos_label would be sufficient.
> 
> Though I would suggest to have ~~~Sensitivity = Recall~~~ and
> 
> ```python
> Class Specificity(Recall):
>     def __init__(self, ..., pos_label=0):
>         ...
> ```

I am completely fine with that. It is very similar to how we have subclass `FBeta` into `F1` because it is a very commonly used metric.

**Comment by Borda** [comment]
@shijianjian mind resolve conflicts so we can proceed further? 🐰

**Comment by shijianjian** [comment]
No problem. Will do it this week.

**Comment by tchaton** [comment]
Hey @shijianjian,

Would you mind resolving conflicts ?

Best,
TC

**Comment by SkafteNicki** [comment]
@shijianjian how is it going here? :]

**Comment by shijianjian** [comment]
Hi guys, sorry for the delay. We are busy finalizing lots of stuff due to the Chinese New Year. Also, those conflicts are introducing deprecations with some behaviour changes. It is not an easy merging. I will rewrite this part according to the new changes by 10th Feb (holiday period).

**Comment by s-rog** [comment]
Will prob need a rebase to feature branch as well? @Borda

**Comment by noamsgl** [comment]
Thank you for this valuable feature

**Comment by Borda** [comment]
Hi @shijianjian, we are very happy that you decided to improve PL and opened this PR.
Never the less we have moved the metric package outside PT, to a separate repository - [torchmetrics](https://github.com/PyTorchLightning/metrics), and unfortunately, there is not GH option to transfer your opened PR there to finish the job there. So I would like to kindly ask you if you are able to finish this PR by end of this week and then we can transfer your merged commit to the new repo or please open the very same PR in the new repository... 🐰 

Anyway thank you very much for being with us ⚡

**Comment by shijianjian** [comment]
It has already been finished at least three times, but conflicts are coming up very quickly ahead of the review. I would say it might be easier to copy and paste the code from this PR to a new PR in the new repo.

**Comment by Borda** [comment]
@shijianjian I fully understand your frustration with fast changes which raises conflicts...
As we already started the deprecation here in PL it would be even more difficult to finish adding a new metric here, may I kindly ask you to reopen the very same PR in `torchmetrics` (the reason is to preserve you as the contribution author) and I promise to help you to finish it there :]

So pls, open PR in TM with the very same content and refer to this PL PR 🐰
